core/src/main/java/hudson/model/ListView.java

```diff
@@ -211,28 +212,49 @@ public boolean hasJobFilterExtensions() {
         }
 
         ItemGroup<? extends TopLevelItem> parent = getOwner().getItemGroup();
-        List<TopLevelItem> parentItems = new ArrayList<>(parent.getItems());
-        includeItems(parent, parentItems, names);
 
         Boolean statusFilter = this.statusFilter; // capture the value to isolate us from concurrent update
-        Iterable<? extends TopLevelItem> candidates;
+        Collection<ViewJobFilter> jobFilters = getJobFilters();
+        List<TopLevelItem> allItems = null;
         if (recurse) {
-            candidates = parent.getAllItems(TopLevelItem.class);
+            allItems = parent.getAllItems(TopLevelItem.class);
+            for (TopLevelItem item : allItems) {
+                String itemName = item.getRelativeNameFrom(parent);
+                if (names.contains(itemName)) {
+                    checkAddItem(statusFilter, items, item);
+                }
+                if (includePattern != null) {
+                    if (includePattern.matcher(itemName).matches()) {
+                        checkAddItem(statusFilter, items, item);
+                    }
+                }
+            }
         } else {
-            candidates = parentItems;
-        }
-        for (TopLevelItem item : candidates) {
-            if (!names.contains(item.getRelativeNameFrom(getOwner().getItemGroup()))) continue;
-            // Add if no status filter or filter matches enabled/disabled status:
-            if(statusFilter == null || !(item instanceof ParameterizedJobMixIn.ParameterizedJob) // TODO or better to call the more generic Job.isBuildable?
-                              || ((ParameterizedJobMixIn.ParameterizedJob)item).isDisabled() ^ statusFilter)
-                items.add(item);
+            for (String name : names) {
+                try {
+                    TopLevelItem i = parent.getItem(name);
+                    if (i != null) {
+                        checkAddItem(statusFilter, items, i);
+                    }
+                } catch (AccessDeniedException e) {
+                    //Ignore
+                }
+            }
+            //Only call getItems if there is a pattern or we have Job Filters
```

Comment:
I think it would help readability if this comment had 2 specific lines of matching code, rather than it explaining the if/else if structure with a repeated statement in block blocks.